### PR TITLE
Fix bad dSYM for symbol-stripped dynamic framework builds

### LINKED ISSUE #6502: Unable to symbolicate crashes from Mapbox 3.3.3 dynamic framework release
When downloading the "mapbox-ios-sdk-3.3.3-dynamic.zip" release attachment on [Mapbox 3.3.3](https://github.com/mapbox/mapbox-gl-native/releases/tag/ios-v3.3.3), the UUIDs from the dSYM do not match the UUIDs from the dynamic framework:

```
$ dwarfdump --uuid mapbox-ios-sdk-3/dynamic/Mapbox.framework/mapbox
UUID: FF7BEA5C-FCB4-3035-A1DF-83707E41ACF1 (i386) mapbox-ios-sdk-3/dynamic/Mapbox.framework/mapbox
UUID: 1E2D5B1D-12AB-3286-8A4D-58B2D417B31C (x86_64) mapbox-ios-sdk-3/dynamic/Mapbox.framework/mapbox
UUID: 71EB9FB8-C32E-3C21-9684-CE53AAAFCB95 (armv7) mapbox-ios-sdk-3/dynamic/Mapbox.framework/mapbox
UUID: 85CF2FD7-622D-3201-90CE-22C24825C0C1 (arm64) mapbox-ios-sdk-3/dynamic/Mapbox.framework/mapbox
```

```
dwarfdump --uuid mapbox-ios-sdk-3/dynamic/Mapbox.framework.dSYM/ 
UUID: E7CE30F1-7197-3155-9A0A-D384EA79502E (armv7) mapbox-ios-sdk-3/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox
UUID: 8CADAD54-2FA2-31E7-B73C-81EFC3B55F7D (arm64) mapbox-ios-sdk-3/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox
```

I've looked through the other release attachments on 3.3.3, and I'm unable to find any dSYM with the `85CF2FD7-622D-3201-90CE-22C24825C0C1`/`71EB9FB8-C32E-3C21-9684-CE53AAAFCB95` UUIDs.

Unfortunately, because of this, we're unable to report ~7 unique crashes that we're receiving in 3.3.3, since we can't symbolicate them.

### PR DESCRIPTION
**Pull request by friedbunny**
Fixes #6502. Stripped dynamic framework packages should now come with the correct dSYM.

Disabling `GCC_GENERATE_DEBUGGING_SYMBOLS` for `SYMBOLS=NO` builds meant that those builds had no debug symbols to strip or add to a dSYM. Running `make iframework SYMBOLS=NO` would fail to create a dSYM, but our deploy script would evidently [reuse the symbol-ful dSYM](https://github.com/mapbox/mapbox-gl-native/issues/6502#issuecomment-250605473).

With the added UUID validation, the above scenario would now fail:
```
* Stripping symbols from binaries…
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: changes being made to the file will invalidate the code signature in: /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/pkg/dynamic/Mapbox.framework/Mapbox (for architecture i386)
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip: changes being made to the file will invalidate the code signature in: /Users/jason/Documents/Mapbox/mapbox-gl-native/build/ios/pkg/dynamic/Mapbox.framework/Mapbox (for architecture x86_64)
* Validating dSYM and framework UUIDs
error: unable to open 'build/ios/pkg/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox': No such file or directory
Error: dSYM and framework UUIDs do not match.
build/ios/pkg/dynamic/Mapbox.framework.dSYM/Contents/Resources/DWARF/Mapbox
  
build/ios/pkg/dynamic/Mapbox.framework/Mapbox
  23F302DF-17AB-384C-879B-746C8A73511A (arm64) 2877864D-E317-3250-8756-11D94C868177 (i386) 59352B56-89D5-3E8E-B83E-9B15657B5C83 (armv7) F91C2CAC-5283-3ADF-ACFB-DD86B4BF5BFF (x86_64)
make: *** [iframework] Error 1
```

Note: `GCC_GENERATE_DEBUGGING_SYMBOLS=YES` is the default setting.

### Useful resources
- https://developer.apple.com/library/content/technotes/tn2004/tn2123.html
- `nm Mapbox.framework/Mapbox | grep MGL` helped test symbol stripping.
- Archived [Xcode build settings reference](https://web.archive.org/web/20160424060222/https://developer.apple.com/library/ios/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html#//apple_ref/doc/uid/TP40003931-CH3-SW53).

/cc @boundsj @1ec5

### CODE REVIEW

**Comment by friedbunny** [comment]
One other thing: the `strip` command was never being invoked because of [bash boolean problems](https://github.com/mapbox/mapbox-gl-native/pull/6531/files#diff-86a1df829aa488a1c4289ab852eb2961L182). Our symbol “stripping” was coming entirely from the fact that we weren’t generating debug symbols.

Now that we _are_ invoking `strip -Sx`, we’re also stripping non-global symbols. So far, this looks like it reduces the raw binary size of Mapbox.framework/Mapbox by 20% — v3.4.0-alpha.4 was 45.3MB, this PR is 35.1MB.

The dSYM size is also much smaller: 234MB vs 32MB. This worries me and I’m not sure why it’s happened, but I was able to successfully symbolicate crashes with this smaller dSYM...

File: platform/macos/scripts/package.sh [hunk 1]
```diff
@@ -43,7 +42,7 @@ if [[ -e ${PRODUCTS}/${BUILDTYPE}/${NAME}.framework.dSYM ]]; then
     cp -r ${PRODUCTS}/${BUILDTYPE}/${NAME}.framework.dSYM "${OUTPUT}"
 fi
 
-if [[ "${GCC_GENERATE_DEBUGGING_SYMBOLS}" == false ]]; then
+if [[ ${SYMBOLS} = NO ]]; then
     step "Stripping binaries…"
     strip -Sx "${OUTPUT}/${NAME}.framework/${NAME}"
```
Comment (1ec5) [inline platform/macos/scripts/package.sh]: Would it be possible to do dSYM validation for macOS as well? I think it might’ve caught #5818 indirectly.

Comment (friedbunny) [inline platform/macos/scripts/package.sh]: Should be as easy as copying it over, I think. Will give it a try.

**Comment by friedbunny** [comment]
- Added macOS dSYM validation.
- Added macOS packaging documentation.
- Went back and ensured that `make iframework BUILDTYPE=Debug` would skip dSYM validation, because there are no dSYMs generated (at least, not until #6529).

Should be ready to merge.

File: platform/macos/scripts/package.sh (hunk 1 above)
Comment (friedbunny) [inline platform/macos/scripts/package.sh]: Did this in 6723a611b02528b8474383c0f91e326e8839b4c3.

**Comment by 1ec5** [comment]
Oh, this is probably worth mentioning in the iOS changelog too.

**Pull request merged**